flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconcilerUpgradeModeTest.java

```diff
@@ -260,11 +265,96 @@ private FlinkDeployment cloneDeploymentWithUpgradeMode(
     }
 
     @ParameterizedTest
-    @EnumSource(UpgradeMode.class)
-    public void testUpgradeBeforeReachingStableSpec(UpgradeMode upgradeMode) throws Exception {
+    @MethodSource("testUpgradeJmDeployCannotStartParams")
+    public void testUpgradeJmDeployCannotStart(UpgradeMode fromMode, UpgradeMode toMode)
+            throws Exception {
```

Comment:
Does it makes sense Gyula to break up this test into multiple methods? It's really hard to decode what are the internal status changes / checks mean actually.
- succesfulJobSubmit
- suspendWithSavepoint
- etc..

Reply:
This test is indeed a bit too complex , I will try to make some extra methods comments to make it clearer

Reply:
I have added some more comments to the tests @morhidi . I could not find a good way to break it up to methods, I feel that would just add more complexity.

These tests are pretty generic in the sense that they test different upgradeMode/savepoint setting combinations. While that helps us to cover all cases in a robust way it makes it a little difficult to simplify beyond a point.